feat: Add support for speculators Eagle checkpoints

rahul-tuli opened the pull request:
# Add support for Eagle models in speculators format

This PR adds support for Eagle models distributed in the "speculators" format, enabling seamless speculative decoding with simple `vllm serve` commands.

## What is speculators format?

Speculators format is an alternative packaging for Eagle models that includes both the draft model and configuration in a single repository. This format simplifies deployment by bundling everything needed for speculative decoding.

## Key Features

### 1. Auto-detection of speculators format
Models in speculators format are automatically detected and configured for speculative decoding:

```bash
# Automatically detects speculators format and configures Eagle speculative decoding
vllm serve nm-testing/eagle-llama3.1-8b-instruct-converted-0717
```

No need for complex `--speculative-config` arguments\!

### 2. Support for multiple Eagle variants

#### Standard Eagle-1 models:
```bash
vllm serve nm-testing/eagle-llama3.1-8b-instruct-converted-0717
```

#### HASS variant (Eagle with additional layernorms):
```bash
# Note: Use VLLM_DISABLE_COMPILE_CACHE=1 to avoid torch.compile cache conflicts
VLLM_DISABLE_COMPILE_CACHE=1 vllm serve nm-testing/hass-llama3.1-8b-layernorms
```

#### Eagle-3 models:
```bash
vllm serve nm-testing/eagle3-llama3.1-8b-instruct-speculators
```

### 3. Qwen model support with Eagle-3
```bash
# Eagle-3 now supports Qwen models
vllm serve nm-testing/Speculator-Qwen3-8B-Eagle3-converted-0717
```

## Example Commands

### Basic usage (auto-detection):
```bash
# The model is automatically detected as speculators format
# Target model, draft model, and speculative config are all configured automatically
vllm serve <speculators-format-model>
```

### With custom tensor parallelism for draft model:
```bash
vllm serve <speculators-format-model> --draft-tensor-parallel-size 1
```

### Manual configuration (if needed):
```bash
# You can still use manual configuration if preferred
vllm serve meta-llama/Llama-3.1-8B-Instruct \
  --speculative-config '{"method": "eagle", "model": "nm-testing/eagle-llama3.1-8b-instruct-converted-0717", "num_speculative_tokens": 5}'
```

## Technical Details

- Automatic detection of speculators format via config structure
- Proper weight remapping between speculators and vLLM naming conventions
- Support for both Eagle-1 and Eagle-3 architectures
- Backwards compatibility with existing Eagle models
- Unified architecture for HASS variants to ensure torch.compile compatibility

## Important Notes

1. **CUDA Graph Caching**: When switching between different Eagle variants (standard vs HASS), use `VLLM_DISABLE_COMPILE_CACHE=1` to avoid torch.compile cache conflicts.

2. **Draft Tensor Parallelism**: Currently, `draft_tensor_parallel_size > 1` is not supported.

3. **V1 Engine**: All examples assume V1 engine usage with `VLLM_USE_V1=1`.

## Testing

Example models for testing:
- Eagle-1: `nm-testing/eagle-llama3.1-8b-instruct-converted-0717`
- HASS: `nm-testing/hass-llama3.1-8b-layernorms`
- Eagle-3: `nm-testing/eagle3-llama3.1-8b-instruct-speculators`
- Qwen + Eagle-3: `nm-testing/Speculator-Qwen3-8B-Eagle3-converted-0717`

dsikka reviewed (commented):
Can you post steps to run?

I wasn't able to run the verification script on this branch with speculators main.

rahul-tuli commented:
> Can you post steps to run?
> 
> I wasn't able to run the verification script on this branch with speculators main.

We don't need speculators to run the models; here are the steps:

Step 1: converted an existing eagle or has checkpoint with speculators convert utility from this branch: https://github.com/neuralmagic/speculators/pull/39 

There is a doc explaining how to use the convert utility here: https://github.com/neuralmagic/speculators/blob/efab1758d803e03f42c85cc67425cefa80c5344f/docs/convert.md

for example convert an existing eagle-checkpoint using:

```bash
speculators convert --eagle yuhuili/EAGLE-LLaMA3.1-Instruct-8B ./converted/eagle meta-llama/Llama-3.1-8B-Instruct
```

Output:
```
2025-07-04 17:14:28.581 | INFO     | speculators.convert.eagle.eagle_converter:convert:50 - Converting Eagle checkpoint: yuhuili/EAGLE-LLaMA3.1-Instruct-8B
2025-07-04 17:14:28.581 | INFO     | speculators.convert.eagle.eagle_converter:_ensure_local:93 - Downloading checkpoint from HuggingFace: yuhuili/EAGLE-LLaMA3.1-Instruct-8B
Fetching 2 files: 100%|█████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00, 31775.03it/s]
2025-07-04 17:14:28.671 | DEBUG    | speculators.convert.eagle.eagle_converter:_ensure_local:100 - Downloaded to: /home/rahul/.cache/huggingface/hub/models--yuhuili--EAGLE-LLaMA3.1-Instruct-8B/snapshots/89073acba22a03994aee0c76774a10ca941e4706
2025-07-04 17:14:28.672 | DEBUG    | speculators.convert.eagle.eagle_converter:_load_checkpoint:118 - Loading config from: /home/rahul/.cache/huggingface/hub/models--yuhuili--EAGLE-LLaMA3.1-Instruct-8B/snapshots/89073acba22a03994aee0c76774a10ca941e4706/config.json
2025-07-04 17:14:28.672 | DEBUG    | speculators.convert.eagle.eagle_converter:_load_checkpoint:133 - Loading PyTorch weights from: /home/rahul/.cache/huggingface/hub/models--yuhuili--EAGLE-LLaMA3.1-Instruct-8B/snapshots/89073acba22a03994aee0c76774a10ca941e4706/pytorch_model.bin
2025-07-04 17:14:29.640 | INFO     | speculators.convert.eagle.eagle_converter:convert:55 - Loaded 10 weights
2025-07-04 17:14:29.640 | DEBUG    | speculators.convert.eagle.eagle_converter:_build_config:169 - Building EagleSpeculatorConfig
2025-07-04 17:14:29.641 | DEBUG    | speculators.convert.eagle.eagle_converter:_build_config:219 - Config built with fusion_bias=False, layernorms=False
2025-07-04 17:14:29.644 | DEBUG    | speculators.convert.eagle.eagle_converter:_process_weights:242 - Processing 10 weights
2025-07-04 17:14:29.644 | DEBUG    | speculators.convert.eagle.eagle_converter:_process_single_weight:277 - Skipping embed_tokens.weight (tied to lm_head)
2025-07-04 17:14:29.644 | DEBUG    | speculators.convert.eagle.eagle_converter:_process_weights:259 - Skipped weights: ['embed_tokens.weight']
2025-07-04 17:14:29.644 | DEBUG    | speculators.convert.eagle.eagle_converter:_process_weights:261 - Remapped weights: ['layers.0.self_attn.q_proj.weight -> transformer.self_attn.q_proj.weight', 'layers.0.self_attn.k_proj.weight -> transformer.self_attn.k_proj.weight', 'layers.0.self_attn.v_proj.weight -> transformer.self_attn.v_proj.weight', 'layers.0.self_attn.o_proj.weight -> transformer.self_attn.o_proj.weight', 'layers.0.mlp.gate_proj.weight -> transformer.mlp.gate_proj.weight', 'layers.0.mlp.up_proj.weight -> transformer.mlp.up_proj.weight', 'layers.0.mlp.down_proj.weight -> transformer.mlp.down_proj.weight', 'layers.0.post_attention_layernorm.weight -> transformer.post_attention_layernorm.weight', 'fc.weight -> fusion_fc.weight']
2025-07-04 17:14:29.700 | DEBUG    | speculators.convert.eagle.eagle_converter:_save_checkpoint:319 - Saving config to: converted/eagle/config.json
2025-07-04 17:14:29.701 | DEBUG    | speculators.convert.eagle.eagle_converter:_save_checkpoint:325 - Saving weights to: converted/eagle/model.safetensors
2025-07-04 17:14:30.157 | SUCCESS  | speculators.convert.eagle.eagle_converter:convert:72 - Saved to: converted/eagle
```

The converted checkpoint will be saved in `converted/eagle` (after this point speculators is not needed)

Step 2: Checkout the current branch in vllm, and run the model:

```python
from vllm import LLM, SamplingParams

# UPDATE THIS PATH
eagle_model_path = "/home/rahul/speculators/converted/eagle"

print("Loading models...")

# Create LLM with Eagle speculative decoding
llm = LLM(
    model="meta-llama/Meta-Llama-3.1-8B-Instruct",  # target/verifier model
    speculative_config={
        "model": eagle_model_path,  # Your Eagle model path
        "num_speculative_tokens": 5,  # Number of tokens to predict ahead
    },
    trust_remote_code=True,
    gpu_memory_utilization=0.4,
    max_model_len=1024, 
)

print("Models loaded! Generating text...")

# Your prompt
prompt = "The benefits of open source software include"


sampling_params = SamplingParams(
    temperature=0.0,
    max_tokens=100,
)

# Generate text
output = llm.generate([prompt], sampling_params)[0]
generated_text = output.outputs[0].text

print(f"\nPrompt: {prompt}")
print(f"Generated: {generated_text}")
```

Output:
```bash
...(truncated for brevity)...
Models loaded! Generating text...
Adding requests: 100%|█████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 74.57it/s]
Processed prompts: 100%|██████████████████████████| 1/1 [00:02<00:00,  2.42s/it, est. speed input: 3.31 toks/s, output: 41.41 toks/s]

Prompt: The benefits of open source software include
Generated: :
1. Cost savings: Open source software is often free or low-cost, which can be a significant advantage for individuals and organizations with limited budgets.
2. Customization: Open source software can be modified and customized to meet specific needs, which can be particularly useful for businesses or organizations with unique requirements.
3. Community support: Open source software often has a large and active community of developers and users who contribute to its development and provide support.
4. Security: Open source software can be more secure
[rank0]:[W704 17:23:14.420996530 ProcessGroupNCCL.cpp:1476] Warning: WARNING: destroy_process_group() was not called before program exit, which can leak resources. For more info, please see https://pytorch.org/docs/stable/distributed.html#shutdown (function operator())
```

dsikka commented:
> > Can you post steps to run?
> > I wasn't able to run the verification script on this branch with speculators main.
> 
> We don't need speculators to run the models; here are the steps:
> 
> Step 1: convert an existing eagle or hass checkpoint with speculators convert utility from this branch (yet to land on main): [neuralmagic/speculators#39](https://github.com/neuralmagic/speculators/pull/39)
> 
> There is a doc explaining how to use the convert utility here: https://github.com/neuralmagic/speculators/blob/efab1758d803e03f42c85cc67425cefa80c5344f/docs/convert.md
> 
> for example convert an existing eagle-checkpoint using:
> 
> ```shell
> speculators convert --eagle yuhuili/EAGLE-LLaMA3.1-Instruct-8B ./converted/eagle meta-llama/Llama-3.1-8B-Instruct
> ```
> 
> Output:
> 
> ```
> 2025-07-04 17:14:28.581 | INFO     | speculators.convert.eagle.eagle_converter:convert:50 - Converting Eagle checkpoint: yuhuili/EAGLE-LLaMA3.1-Instruct-8B
> 2025-07-04 17:14:28.581 | INFO     | speculators.convert.eagle.eagle_converter:_ensure_local:93 - Downloading checkpoint from HuggingFace: yuhuili/EAGLE-LLaMA3.1-Instruct-8B
> Fetching 2 files: 100%|█████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00, 31775.03it/s]
> 2025-07-04 17:14:28.671 | DEBUG    | speculators.convert.eagle.eagle_converter:_ensure_local:100 - Downloaded to: /home/rahul/.cache/huggingface/hub/models--yuhuili--EAGLE-LLaMA3.1-Instruct-8B/snapshots/89073acba22a03994aee0c76774a10ca941e4706
> 2025-07-04 17:14:28.672 | DEBUG    | speculators.convert.eagle.eagle_converter:_load_checkpoint:118 - Loading config from: /home/rahul/.cache/huggingface/hub/models--yuhuili--EAGLE-LLaMA3.1-Instruct-8B/snapshots/89073acba22a03994aee0c76774a10ca941e4706/config.json
> 2025-07-04 17:14:28.672 | DEBUG    | speculators.convert.eagle.eagle_converter:_load_checkpoint:133 - Loading PyTorch weights from: /home/rahul/.cache/huggingface/hub/models--yuhuili--EAGLE-LLaMA3.1-Instruct-8B/snapshots/89073acba22a03994aee0c76774a10ca941e4706/pytorch_model.bin
> 2025-07-04 17:14:29.640 | INFO     | speculators.convert.eagle.eagle_converter:convert:55 - Loaded 10 weights
> 2025-07-04 17:14:29.640 | DEBUG    | speculators.convert.eagle.eagle_converter:_build_config:169 - Building EagleSpeculatorConfig
> 2025-07-04 17:14:29.641 | DEBUG    | speculators.convert.eagle.eagle_converter:_build_config:219 - Config built with fusion_bias=False, layernorms=False
> 2025-07-04 17:14:29.644 | DEBUG    | speculators.convert.eagle.eagle_converter:_process_weights:242 - Processing 10 weights
> 2025-07-04 17:14:29.644 | DEBUG    | speculators.convert.eagle.eagle_converter:_process_single_weight:277 - Skipping embed_tokens.weight (tied to lm_head)
> 2025-07-04 17:14:29.644 | DEBUG    | speculators.convert.eagle.eagle_converter:_process_weights:259 - Skipped weights: ['embed_tokens.weight']
> 2025-07-04 17:14:29.644 | DEBUG    | speculators.convert.eagle.eagle_converter:_process_weights:261 - Remapped weights: ['layers.0.self_attn.q_proj.weight -> transformer.self_attn.q_proj.weight', 'layers.0.self_attn.k_proj.weight -> transformer.self_attn.k_proj.weight', 'layers.0.self_attn.v_proj.weight -> transformer.self_attn.v_proj.weight', 'layers.0.self_attn.o_proj.weight -> transformer.self_attn.o_proj.weight', 'layers.0.mlp.gate_proj.weight -> transformer.mlp.gate_proj.weight', 'layers.0.mlp.up_proj.weight -> transformer.mlp.up_proj.weight', 'layers.0.mlp.down_proj.weight -> transformer.mlp.down_proj.weight', 'layers.0.post_attention_layernorm.weight -> transformer.post_attention_layernorm.weight', 'fc.weight -> fusion_fc.weight']
> 2025-07-04 17:14:29.700 | DEBUG    | speculators.convert.eagle.eagle_converter:_save_checkpoint:319 - Saving config to: converted/eagle/config.json
> 2025-07-04 17:14:29.701 | DEBUG    | speculators.convert.eagle.eagle_converter:_save_checkpoint:325 - Saving weights to: converted/eagle/model.safetensors
> 2025-07-04 17:14:30.157 | SUCCESS  | speculators.convert.eagle.eagle_converter:convert:72 - Saved to: converted/eagle
> ```
> 
> The converted checkpoint will be saved in `converted/eagle` (after this point speculators is not needed)
> 
> Step 2: Checkout the current branch in vllm, and run the model:
> 
> ```python
> from vllm import LLM, SamplingParams
> 
> # UPDATE THIS PATH
> eagle_model_path = "/home/rahul/speculators/converted/eagle"
> 
> print("Loading models...")
> 
> # Create LLM with Eagle speculative decoding
> llm = LLM(
>     model="meta-llama/Meta-Llama-3.1-8B-Instruct",  # target/verifier model
>     speculative_config={
>         "model": eagle_model_path,  # Your Eagle model path
>         "num_speculative_tokens": 5,  # Number of tokens to predict ahead
>     },
>     trust_remote_code=True,
>     gpu_memory_utilization=0.4,
>     max_model_len=1024, 
> )
> 
> print("Models loaded! Generating text...")
> 
> # Your prompt
> prompt = "The benefits of open source software include"
> 
> 
> sampling_params = SamplingParams(
>     temperature=0.0,
>     max_tokens=100,
> )
> 
> # Generate text
> output = llm.generate([prompt], sampling_params)[0]
> generated_text = output.outputs[0].text
> 
> print(f"\nPrompt: {prompt}")
> print(f"Generated: {generated_text}")
> ```
> 
> Output:
> 
> ```shell
> ...(truncated for brevity)...
> Models loaded! Generating text...
> Adding requests: 100%|█████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 74.57it/s]
> Processed prompts: 100%|██████████████████████████| 1/1 [00:02<00:00,  2.42s/it, est. speed input: 3.31 toks/s, output: 41.41 toks/s]
> 
> Prompt: The benefits of open source software include
> Generated: :
> 1. Cost savings: Open source software is often free or low-cost, which can be a significant advantage for individuals and organizations with limited budgets.
> 2. Customization: Open source software can be modified and customized to meet specific needs, which can be particularly useful for businesses or organizations with unique requirements.
> 3. Community support: Open source software often has a large and active community of developers and users who contribute to its development and provide support.
> 4. Security: Open source software can be more secure
> [rank0]:[W704 17:23:14.420996530 ProcessGroupNCCL.cpp:1476] Warning: WARNING: destroy_process_group() was not called before program exit, which can leak resources. For more info, please see https://pytorch.org/docs/stable/distributed.html#shutdown (function operator())
> ```

cli does not seem to be working on that branch

rahul-tuli commented:
> > > Can you post steps to run?
> 
> > > I wasn't able to run the verification script on this branch with speculators main.
> 
> > 
> 
> > We don't need speculators to run the models; here are the steps:
> 
> > 
> 
> > Step 1: convert an existing eagle or hass checkpoint with speculators convert utility from this branch (yet to land on main): [neuralmagic/speculators#39](https://github.com/neuralmagic/speculators/pull/39)
> 
> > 
> 
> > There is a doc explaining how to use the convert utility here: https://github.com/neuralmagic/speculators/blob/efab1758d803e03f42c85cc67425cefa80c5344f/docs/convert.md
> 
> > 
> 
> > for example convert an existing eagle-checkpoint using:
> 
> > 
> 
> > ```shell
> 
> > speculators convert --eagle yuhuili/EAGLE-LLaMA3.1-Instruct-8B ./converted/eagle meta-llama/Llama-3.1-8B-Instruct
> 
> > ```
> 
> > 
> 
> > Output:
> 
> > 
> 
> > ```
> 
> > 2025-07-04 17:14:28.581 | INFO     | speculators.convert.eagle.eagle_converter:convert:50 - Converting Eagle checkpoint: yuhuili/EAGLE-LLaMA3.1-Instruct-8B
> 
> > 2025-07-04 17:14:28.581 | INFO     | speculators.convert.eagle.eagle_converter:_ensure_local:93 - Downloading checkpoint from HuggingFace: yuhuili/EAGLE-LLaMA3.1-Instruct-8B
> 
> > Fetching 2 files: 100%|█████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00, 31775.03it/s]
> 
> > 2025-07-04 17:14:28.671 | DEBUG    | speculators.convert.eagle.eagle_converter:_ensure_local:100 - Downloaded to: /home/rahul/.cache/huggingface/hub/models--yuhuili--EAGLE-LLaMA3.1-Instruct-8B/snapshots/89073acba22a03994aee0c76774a10ca941e4706
> 
> > 2025-07-04 17:14:28.672 | DEBUG    | speculators.convert.eagle.eagle_converter:_load_checkpoint:118 - Loading config from: /home/rahul/.cache/huggingface/hub/models--yuhuili--EAGLE-LLaMA3.1-Instruct-8B/snapshots/89073acba22a03994aee0c76774a10ca941e4706/config.json
> 
> > 2025-07-04 17:14:28.672 | DEBUG    | speculators.convert.eagle.eagle_converter:_load_checkpoint:133 - Loading PyTorch weights from: /home/rahul/.cache/huggingface/hub/models--yuhuili--EAGLE-LLaMA3.1-Instruct-8B/snapshots/89073acba22a03994aee0c76774a10ca941e4706/pytorch_model.bin
> 
> > 2025-07-04 17:14:29.640 | INFO     | speculators.convert.eagle.eagle_converter:convert:55 - Loaded 10 weights
> 
> > 2025-07-04 17:14:29.640 | DEBUG    | speculators.convert.eagle.eagle_converter:_build_config:169 - Building EagleSpeculatorConfig
> 
> > 2025-07-04 17:14:29.641 | DEBUG    | speculators.convert.eagle.eagle_converter:_build_config:219 - Config built with fusion_bias=False, layernorms=False
> 
> > 2025-07-04 17:14:29.644 | DEBUG    | speculators.convert.eagle.eagle_converter:_process_weights:242 - Processing 10 weights
> 
> > 2025-07-04 17:14:29.644 | DEBUG    | speculators.convert.eagle.eagle_converter:_process_single_weight:277 - Skipping embed_tokens.weight (tied to lm_head)
> 
> > 2025-07-04 17:14:29.644 | DEBUG    | speculators.convert.eagle.eagle_converter:_process_weights:259 - Skipped weights: ['embed_tokens.weight']
> 
> > 2025-07-04 17:14:29.644 | DEBUG    | speculators.convert.eagle.eagle_converter:_process_weights:261 - Remapped weights: ['layers.0.self_attn.q_proj.weight -> transformer.self_attn.q_proj.weight', 'layers.0.self_attn.k_proj.weight -> transformer.self_attn.k_proj.weight', 'layers.0.self_attn.v_proj.weight -> transformer.self_attn.v_proj.weight', 'layers.0.self_attn.o_proj.weight -> transformer.self_attn.o_proj.weight', 'layers.0.mlp.gate_proj.weight -> transformer.mlp.gate_proj.weight', 'layers.0.mlp.up_proj.weight -> transformer.mlp.up_proj.weight', 'layers.0.mlp.down_proj.weight -> transformer.mlp.down_proj.weight', 'layers.0.post_attention_layernorm.weight -> transformer.post_attention_layernorm.weight', 'fc.weight -> fusion_fc.weight']
> 
> > 2025-07-04 17:14:29.700 | DEBUG    | speculators.convert.eagle.eagle_converter:_save_checkpoint:319 - Saving config to: converted/eagle/config.json
> 
> > 2025-07-04 17:14:29.701 | DEBUG    | speculators.convert.eagle.eagle_converter:_save_checkpoint:325 - Saving weights to: converted/eagle/model.safetensors
> 
> > 2025-07-04 17:14:30.157 | SUCCESS  | speculators.convert.eagle.eagle_converter:convert:72 - Saved to: converted/eagle
> 
> > ```
> 
> > 
> 
> > The converted checkpoint will be saved in `converted/eagle` (after this point speculators is not needed)
> 
> > 
> 
> > Step 2: Checkout the current branch in vllm, and run the model:
> 
> > 
> 
> > ```python
> 
> > from vllm import LLM, SamplingParams
> 
> > 
> 
> > # UPDATE THIS PATH
> 
> > eagle_model_path = "/home/rahul/speculators/converted/eagle"
> 
> > 
> 
> > print("Loading models...")
> 
> > 
> 
> > # Create LLM with Eagle speculative decoding
> 
> > llm = LLM(
> 
> >     model="meta-llama/Meta-Llama-3.1-8B-Instruct",  # target/verifier model
> 
> >     speculative_config={
> 
> >         "model": eagle_model_path,  # Your Eagle model path
> 
> >         "num_speculative_tokens": 5,  # Number of tokens to predict ahead
> 
> >     },
> 
> >     trust_remote_code=True,
> 
> >     gpu_memory_utilization=0.4,
> 
> >     max_model_len=1024, 
> 
> > )
> 
> > 
> 
> > print("Models loaded! Generating text...")
> 
> > 
> 
> > # Your prompt
> 
> > prompt = "The benefits of open source software include"
> 
> > 
> 
> > 
> 
> > sampling_params = SamplingParams(
> 
> >     temperature=0.0,
> 
> >     max_tokens=100,
> 
> > )
> 
> > 
> 
> > # Generate text
> 
> > output = llm.generate([prompt], sampling_params)[0]
> 
> > generated_text = output.outputs[0].text
> 
> > 
> 
> > print(f"\nPrompt: {prompt}")
> 
> > print(f"Generated: {generated_text}")
> 
> > ```
> 
> > 
> 
> > Output:
> 
> > 
> 
> > ```shell
> 
> > ...(truncated for brevity)...
> 
> > Models loaded! Generating text...
> 
> > Adding requests: 100%|█████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 74.57it/s]
> 
> > Processed prompts: 100%|██████████████████████████| 1/1 [00:02<00:00,  2.42s/it, est. speed input: 3.31 toks/s, output: 41.41 toks/s]
> 
> > 
> 
> > Prompt: The benefits of open source software include
> 
> > Generated: :
> 
> > 1. Cost savings: Open source software is often free or low-cost, which can be a significant advantage for individuals and organizations with limited budgets.
> 
> > 2. Customization: Open source software can be modified and customized to meet specific needs, which can be particularly useful for businesses or organizations with unique requirements.
> 
> > 3. Community support: Open source software often has a large and active community of developers and users who contribute to its development and provide support.
> 
> > 4. Security: Open source software can be more secure
> 
> > [rank0]:[W704 17:23:14.420996530 ProcessGroupNCCL.cpp:1476] Warning: WARNING: destroy_process_group() was not called before program exit, which can leak resources. For more info, please see https://pytorch.org/docs/stable/distributed.html#shutdown (function operator())
> 
> > ```
> 
> 
> 
> cli does not seem to be working on that branch.

Could you paste the command you used, and the trace back on that PR?

dsikka commented:
> > > > Can you post steps to run?
> > 
> > 
> > > > I wasn't able to run the verification script on this branch with speculators main.
> > 
> > 
> > > 
> > 
> > 
> > > We don't need speculators to run the models; here are the steps:
> > 
> > 
> > > 
> > 
> > 
> > > Step 1: convert an existing eagle or hass checkpoint with speculators convert utility from this branch (yet to land on main): [neuralmagic/speculators#39](https://github.com/neuralmagic/speculators/pull/39)
> > 
> > 
> > > 
> > 
> > 
> > > There is a doc explaining how to use the convert utility here: https://github.com/neuralmagic/speculators/blob/efab1758d803e03f42c85cc67425cefa80c5344f/docs/convert.md
> > 
> > 
> > > 
> > 
> > 
> > > for example convert an existing eagle-checkpoint using:
> > 
> > 
> > > 
> > 
> > 
> > > 
> > 
> > 
> > > speculators convert --eagle yuhuili/EAGLE-LLaMA3.1-Instruct-8B ./converted/eagle meta-llama/Llama-3.1-8B-Instruct
> > 
> > 
> > > 
> > 
> > 
> > > 
> > 
> > 
> > > Output:
> > 
> > 
> > > 
> > 
> > 
> > > 
> > 
> > 
> > > 2025-07-04 17:14:28.581 | INFO     | speculators.convert.eagle.eagle_converter:convert:50 - Converting Eagle checkpoint: yuhuili/EAGLE-LLaMA3.1-Instruct-8B
> > 
> > 
> > > 2025-07-04 17:14:28.581 | INFO     | speculators.convert.eagle.eagle_converter:_ensure_local:93 - Downloading checkpoint from HuggingFace: yuhuili/EAGLE-LLaMA3.1-Instruct-8B
> > 
> > 
> > > Fetching 2 files: 100%|█████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00, 31775.03it/s]
> > 
> > 
> > > 2025-07-04 17:14:28.671 | DEBUG    | speculators.convert.eagle.eagle_converter:_ensure_local:100 - Downloaded to: /home/rahul/.cache/huggingface/hub/models--yuhuili--EAGLE-LLaMA3.1-Instruct-8B/snapshots/89073acba22a03994aee0c76774a10ca941e4706
> > 
> > 
> > > 2025-07-04 17:14:28.672 | DEBUG    | speculators.convert.eagle.eagle_converter:_load_checkpoint:118 - Loading config from: /home/rahul/.cache/huggingface/hub/models--yuhuili--EAGLE-LLaMA3.1-Instruct-8B/snapshots/89073acba22a03994aee0c76774a10ca941e4706/config.json
> > 
> > 
> > > 2025-07-04 17:14:28.672 | DEBUG    | speculators.convert.eagle.eagle_converter:_load_checkpoint:133 - Loading PyTorch weights from: /home/rahul/.cache/huggingface/hub/models--yuhuili--EAGLE-LLaMA3.1-Instruct-8B/snapshots/89073acba22a03994aee0c76774a10ca941e4706/pytorch_model.bin
> > 
> > 
> > > 2025-07-04 17:14:29.640 | INFO     | speculators.convert.eagle.eagle_converter:convert:55 - Loaded 10 weights
> > 
> > 
> > > 2025-07-04 17:14:29.640 | DEBUG    | speculators.convert.eagle.eagle_converter:_build_config:169 - Building EagleSpeculatorConfig
> > 
> > 
> > > 2025-07-04 17:14:29.641 | DEBUG    | speculators.convert.eagle.eagle_converter:_build_config:219 - Config built with fusion_bias=False, layernorms=False
> > 
> > 
> > > 2025-07-04 17:14:29.644 | DEBUG    | speculators.convert.eagle.eagle_converter:_process_weights:242 - Processing 10 weights
> > 
> > 
> > > 2025-07-04 17:14:29.644 | DEBUG    | speculators.convert.eagle.eagle_converter:_process_single_weight:277 - Skipping embed_tokens.weight (tied to lm_head)
> > 
> > 
> > > 2025-07-04 17:14:29.644 | DEBUG    | speculators.convert.eagle.eagle_converter:_process_weights:259 - Skipped weights: ['embed_tokens.weight']
> > 
> > 
> > > 2025-07-04 17:14:29.644 | DEBUG    | speculators.convert.eagle.eagle_converter:_process_weights:261 - Remapped weights: ['layers.0.self_attn.q_proj.weight -> transformer.self_attn.q_proj.weight', 'layers.0.self_attn.k_proj.weight -> transformer.self_attn.k_proj.weight', 'layers.0.self_attn.v_proj.weight -> transformer.self_attn.v_proj.weight', 'layers.0.self_attn.o_proj.weight -> transformer.self_attn.o_proj.weight', 'layers.0.mlp.gate_proj.weight -> transformer.mlp.gate_proj.weight', 'layers.0.mlp.up_proj.weight -> transformer.mlp.up_proj.weight', 'layers.0.mlp.down_proj.weight -> transformer.mlp.down_proj.weight', 'layers.0.post_attention_layernorm.weight -> transformer.post_attention_layernorm.weight', 'fc.weight -> fusion_fc.weight']
> > 
> > 
> > > 2025-07-04 17:14:29.700 | DEBUG    | speculators.convert.eagle.eagle_converter:_save_checkpoint:319 - Saving config to: converted/eagle/config.json
> > 
> > 
> > > 2025-07-04 17:14:29.701 | DEBUG    | speculators.convert.eagle.eagle_converter:_save_checkpoint:325 - Saving weights to: converted/eagle/model.safetensors
> > 
> > 
> > > 2025-07-04 17:14:30.157 | SUCCESS  | speculators.convert.eagle.eagle_converter:convert:72 - Saved to: converted/eagle
> > 
> > 
> > > 
> > 
> > 
> > > 
> > 
> > 
> > > The converted checkpoint will be saved in `converted/eagle` (after this point speculators is not needed)
> > 
> > 
> > > 
> > 
> > 
> > > Step 2: Checkout the current branch in vllm, and run the model:
> > 
> > 
> > > 
> > 
> > 
> > > 
> > 
> > 
> > > from vllm import LLM, SamplingParams
> > 
> > 
> > > 
> > 
> > 
> > > # UPDATE THIS PATH
> > 
> > 
> > > eagle_model_path = "/home/rahul/speculators/converted/eagle"
> > 
> > 
> > > 
> > 
> > 
> > > print("Loading models...")
> > 
> > 
> > > 
> > 
> > 
> > > # Create LLM with Eagle speculative decoding
> > 
> > 
> > > llm = LLM(
> > 
> > 
> > > ```
> > > model="meta-llama/Meta-Llama-3.1-8B-Instruct",  # target/verifier model
> > > ```
> > 
> > 
> > > ```
> > > speculative_config={
> > > ```
> > 
> > 
> > > ```
> > >     "model": eagle_model_path,  # Your Eagle model path
> > > ```
> > 
> > 
> > > ```
> > >     "num_speculative_tokens": 5,  # Number of tokens to predict ahead
> > > ```
> > 
> > 
> > > ```
> > > },
> > > ```
> > 
> > 
> > > ```
> > > trust_remote_code=True,
> > > ```
> > 
> > 
> > > ```
> > > gpu_memory_utilization=0.4,
> > > ```
> > 
> > 
> > > ```
> > > max_model_len=1024, 
> > > ```
> > 
> > 
> > > )
> > 
> > 
> > > 
> > 
> > 
> > > print("Models loaded! Generating text...")
> > 
> > 
> > > 
> > 
> > 
> > > # Your prompt
> > 
> > 
> > > prompt = "The benefits of open source software include"
> > 
> > 
> > > 
> > 
> > 
> > > 
> > 
> > 
> > > sampling_params = SamplingParams(
> > 
> > 
> > > ```
> > > temperature=0.0,
> > > ```
> > 
> > 
> > > ```
> > > max_tokens=100,
> > > ```
> > 
> > 
> > > )
> > 
> > 
> > > 
> > 
> > 
> > > # Generate text
> > 
> > 
> > > output = llm.generate([prompt], sampling_params)[0]
> > 
> > 
> > > generated_text = output.outputs[0].text
> > 
> > 
> > > 
> > 
> > 
> > > print(f"\nPrompt: {prompt}")
> > 
> > 
> > > print(f"Generated: {generated_text}")
> > 
> > 
> > > 
> > 
> > 
> > > 
> > 
> > 
> > > Output:
> > 
> > 
> > > 
> > 
> > 
> > > 
> > 
> > 
> > > ...(truncated for brevity)...
> > 
> > 
> > > Models loaded! Generating text...
> > 
> > 
> > > Adding requests: 100%|█████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 74.57it/s]
> > 
> > 
> > > Processed prompts: 100%|██████████████████████████| 1/1 [00:02<00:00,  2.42s/it, est. speed input: 3.31 toks/s, output: 41.41 toks/s]
> > 
> > 
> > > 
> > 
> > 
> > > Prompt: The benefits of open source software include
> > 
> > 
> > > Generated: :
> > 
> > 
> > > 1. Cost savings: Open source software is often free or low-cost, which can be a significant advantage for individuals and organizations with limited budgets.
> > 
> > 
> > > 2. Customization: Open source software can be modified and customized to meet specific needs, which can be particularly useful for businesses or organizations with unique requirements.
> > 
> > 
> > > 3. Community support: Open source software often has a large and active community of developers and users who contribute to its development and provide support.
> > 
> > 
> > > 4. Security: Open source software can be more secure
> > 
> > 
> > > [rank0]:[W704 17:23:14.420996530 ProcessGroupNCCL.cpp:1476] Warning: WARNING: destroy_process_group() was not called before program exit, which can leak resources. For more info, please see https://pytorch.org/docs/stable/distributed.html#shutdown (function operator())
> > 
> > 
> > > 
> > 
> > 
> > cli does not seem to be working on that branch.
> 
> Could you paste the command you used, and the trace back on that PR?

I’ll touch base offline. It just didn’t recognize the speculators command

dsikka commented on vllm/transformers_utils/configs/speculators_eagle.py:
```diff
@@ -0,0 +1,114 @@
+                transformer_config["architectures"] = [arch]
+        
+        # Build vLLM config
+        vllm_config = {
```
Why don't we need to add the verifier model as part of the config?

dsikka reviewed (commented):
question

dsikka commented on vllm/model_executor/models/llama_eagle.py:
```diff
@@ -22,6 +23,23 @@
 
+# Weight name mapping for speculators format compatibility
+SPECULATORS_WEIGHT_MAP = {
+    "fusion_fc.weight": "fc.weight",
```
I though we removed this translation in speculators?

rahul-tuli replied:
we haven't so far, there was contention behind eagle1 naming, we will remove the need for this translation in a follow up PR

dsikka commented on vllm/transformers_utils/configs/speculators_eagle.py:
```diff
@@ -0,0 +1,352 @@
+DEFAULT_NUM_LOOKAHEAD_TOKENS = 5
+
+
+class SpeculatorsEagleConfig(EAGLEConfig):
```
Why did we change this from SpeculatorsEagle config?

rahul-tuli replied:
To avoid confusion with SpeculativeConfig! we can think of a better name

dsikka commented on vllm/config.py:
```diff
@@ -2767,9 +2767,15 @@ def __post_init__(self):
                     pass
+                elif hasattr(self.draft_model_config.hf_config, 
+                           "speculators_model_type") and \
+                        self.draft_model_config.hf_config.speculators_model_type in ("eagle", "eagle3"):
```
why do we need this

dsikka commented on vllm/config.py:
```diff
@@ -2767,9 +2767,15 @@ def __post_init__(self):
                 elif "eagle-" in self.draft_model_config.model.lower() or \
                         "eagle3-" in self.draft_model_config.model.lower():
                     self.method = "eagle"
+                elif self.draft_model_config.hf_config.model_type == "eagle":
```
same as above

dsikka commented on vllm/model_executor/models/llama_eagle.py:
```diff
@@ -22,6 +24,27 @@
 
 logger = init_logger(__name__)
 
+# Map speculators weight names to vLLM names
```
Should probably live in the speculators config file

dsikka commented on vllm/model_executor/models/llama_eagle.py:
```diff
@@ -22,6 +24,27 @@
+}
+
+
+def remap_speculators_weight_name(name: str) -> Optional[str]:
```
same as above

dsikka commented on vllm/transformers_utils/configs/speculators_eagle.py:
```diff
@@ -0,0 +1,284 @@
+        return num_lookahead_tokens
+    
+    @classmethod
+    def _apply_eagle_v1_config(
```
these are fine for now but wondering if we can combine the eagle1/eagle3 _apply methods defined here

rahul-tuli committed:
feat: Add support for Eagle models in speculators format

- Add SpeculatorsEAGLEConfig for handling speculators format Eagle models
- Support both Eagle-1 and Eagle-3 variants with proper config translation
- Add detection function to identify speculators format models
- Handle weight remapping between speculators and vLLM naming conventions

Signed-off-by: Rahul Tuli <rtuli@redhat.com>

rahul-tuli committed:
feat: Update Eagle models for speculators format compatibility

- Add weight remapping function for speculators format in llama_eagle.py
- Support HASS variant with additional layernorms
- Add IdentityNorm for unified architecture to fix torch.compile compatibility
- Propagate norm_before_residual config for Eagle3 models

Signed-off-by: Rahul Tuli <rtuli@redhat.com>

rahul-tuli committed:
feat: Add Eagle3 support for Qwen models

- Enable Eagle3 speculative decoding for Qwen2 and Qwen3
- Import and register Eagle3 in supported draft model types

Signed-off-by: Rahul Tuli <rtuli@redhat.com>

rahul-tuli committed:
feat: Enable auto-detection for speculators format models

- Auto-detect speculators format models and configure speculative decoding
- Allow 'vllm serve <speculators-model>' without explicit speculative config
- Add draft_tensor_parallel_size CLI argument
- Update tokenizer to use target model instead of draft model
- Check speculators model type in draft config for proper Eagle detection

Signed-off-by: Rahul Tuli <rtuli@redhat.com>

rahul-tuli committed:
docs: Add example scripts for serving Eagle models

- Add Eagle1 example scripts for Llama models
- Add Eagle3 example scripts for Llama and Qwen models
- Include both regular and float16 serving examples

Signed-off-by: Rahul Tuli <rtuli@redhat.com>

rahul-tuli committed:
Add: README.nd for eagle1 examples, remove --enforce-eager flag

Signed-off-by: Rahul Tuli <rtuli@redhat.com>

rahul-tuli committed:
Move speculators weight map and remap functions to vllm/transformers_utils/configs/speculators_eagle.py

Signed-off-by: Rahul Tuli <rtuli@redhat.com>

rahul-tuli committed:
Some minor cleanups

Signed-off-by: Rahul Tuli <rtuli@redhat.com>

rahul-tuli committed:
pre-commit hooks

Signed-off-by: Rahul Tuli <rtuli@redhat.com>

rahul-tuli committed:
remove changes to requirements/test.txt

Signed-off-by: Rahul Tuli <rtuli@redhat.com>

dsikka commented on requirements/test.txt:
```diff
@@ -174,8 +174,6 @@ fsspec==2024.9.0
     #   huggingface-hub
     #   torch
-ftfy==6.3.1
-    # via open-clip-torch
```
rebase error?

rahul-tuli replied:
Good catch!

dsikka reviewed (commented):
Overall LGTM. I think we can potentially clean-up additionally by moving some of the logic into the SpeculatorsEagleConfig/SpeculatorsConfig

rahul-tuli committed:
remove file added in faulty rebase

Signed-off-by: Rahul Tuli <rtuli@redhat.com>

rahul-tuli committed:
Fix bug with Qwen

Signed-off-by: Rahul Tuli <rtuli@redhat.com>

dsikka committed:
Merge branch 'main' into feat/speculators-eagle-support
